### PR TITLE
DIP 22: Deterministic InstantSend

### PR DESCRIPTION
**Pull request by thephez**
This DIP aims to improve InstantSend messages to make them deterministically verifiable.

Link to doc: https://github.com/dashpay/dips/blob/deterministic-instantsend/dip-0022.md

Thanks to @QuantumExplorer and @UdjinM6 for collaborating on this

### CODE REVIEW

File: deterministic-instantsend.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+| **cycleHash** | uint256 | 32 | Block hash of first block of the cycle in which the quorum signing this `islock` is active |
+| sig | BLSSig | 96 | Recovered signature from the signing request/session |
+
+### **Choosing the active LLMQ to perform signing**
+
+Choosing the active LLMQ to perform signing should follow the same steps as defined in [DIP-0007 -
+Choosing the active LLMQ to perform
+signing](https://github.com/dashpay/dips/blob/master/dip-0007.md#choosing-the-active-llmq-to-perform-signing).
+
+| Field | Type | Size | Description |
```
Comment (thephez) [inline deterministic-instantsend.md:109]: @QuantumExplorer Is this table supposed to be here? I'm not sure I understand its purpose.

File: deterministic-instantsend.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+known. The drawback of this approach is that any quorum could sign any `islock` even if it were not
+responsible to do so for that quorum cycle.
+
+A quorum cycle begins at a `quorumBlock` (as per
+[DIP-0006](https://github.com/dashpay/dips/blob/master/dip-0006.md#parametersvariables-of-a-llmq-and-dkg))
+and lasts for a number of blocks that is equal to the `quorumDkgInterval`. During this time the set
+of valid quorums are not modified for a given `quorumType`. If the `quorumDkgInterval` is set to 24
+blocks then from block 100 to block 124 quorums of that type stay the same. During this period the
+quorum that is in charge of signing a specific `islock` will also always be the same. CycleHash is
+the `blockHash` of the first block in a cycle.
```
Comment (thephez) [inline deterministic-instantsend.md:55]: Is this the agreed upon definition for cycle hash? There was previously some discussion regarding exactly which block's hash should be used. @QuantumExplorer @UdjinM6

**Review by thephez — COMMENTED** [review]
A couple questions for @QuantumExplorer and @UdjinM6

File: deterministic-instantsend.md (hunk 1 above)
Comment (thephez) [reply deterministic-instantsend.md]: Removed after talking with Pasta

File: deterministic-instantsend.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+By adding the `cycleHash` to the `islock` message, any node can follow the steps required to
+determine the appropriate `quorumHash` and verify the signature.
```
Comment (thephez) [inline deterministic-instantsend.md:58]:
```suggestion
determine the appropriate `quorumHash`. And add something else here...
```

File: deterministic-instantsend.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+* [DIP-0007: LLMQ Signing Requests / Sessions](https://github.com/dashpay/dips/blob/master/dip-0007.md)
+* [DIP-0010: LLMQ InstantSend](https://github.com/dashpay/dips/blob/master/dip-0010.md)
+
+## Versioning of ISLock messages
+
+Since `islock` messages were never versioned, a new `ISDLOCK` message will be created and the
+`islock` message will be deprecated. ISD stands for InstantSend Deterministic. The version of the
+`ISDLOCK` used in this document will be `1`. We will still refer to `islock` messages, even though
+the message name has been changed.
+
```
Comment (thephez) [inline deterministic-instantsend.md:39]:
```suggestion
```

**Review by PastaPastaPasta — APPROVED** [review]

**Pull request merged**